Source/DataSourceProvider.swift

```diff
@@ -3,218 +3,339 @@
+        guard let item = item else {
+            return
+        }
+
+        let numberOfChildren = item.children.count
+        guard numberOfChildren > 0 else {
+            return
+        }
+
+        let selectedParentCell: ParentCell = ParentCell(
+                indexPath: indexPath,
+                index: parentIndex)
+
+        tableView.beginUpdates()
+        toggle(selectedParentCell, withState: item.state, tableView)
+        tableView.endUpdates()
+
+        // If the cells were expanded then we verify if they are inside the CGRect
+        if item.state == .expanded {
+            let lastCellIndexPath = IndexPath(item: indexPath.item + numberOfChildren, section: indexPath.section)
+            // Scroll the new cells expanded in case of be outside the UITableView CGRect
+            scrollCellIfNeeded(atIndexPath: lastCellIndexPath, tableView)
+        }
+    }
+
+// Toggle the state of the selected parent cell between expanded and collapsed
+//
+// - Parameters:
+//   - currentState: The current state of the selected parent
+//   - selectedParentCell: The actual cell selected
+    private func toggle(_ selectedParentCell: ParentCell, withState currentState: State, _ tableView: UITableView) {
+        switch (currentState, numberOfExpandedParentCells) {
+        case (.expanded, _):
+            // Collapse the parent and it's children
+            collapse(parent: selectedParentCell, tableView)
+            expandedParent = nil
+        case (.collapsed, .single):
+            // Expand the parent and it's children and collapse the expanded parent
+            if let expandedParent = expandedParent {
+                collapse(parent: expandedParent, tableView)
+            }
+            expand(parent: selectedParentCell, tableView)
+            expandedParent = selectedParentCell
+        case (.collapsed, .multiple):
+            // Expand the parent and it's children
+            expand(parent: selectedParentCell, tableView)
+        }
+    }
+
+// Expand the parent cell and it's children
+//
+// - Parameters:
+//   - parent: The actual parent cell to be expanded
+    private func expand(parent: ParentCell, _ tableView: UITableView) {
+        let numberOfChildren = dataSource.item(atRow: parent.index, inSection: parent.indexPath.section)?.children.count ?? 0
+
+        guard numberOfChildren > 0 else {
+            return
+        }
+
+        let indexPaths = getIndexes(parent, numberOfChildren)
+        tableView.insertRows(at: indexPaths, with: .fade)
+        dataSource.toggleParentCell(toState: .expanded, inSection: parent.indexPath.section, atIndex: parent.index)
+    }
+
+// Collapse the parent cell and it's children
+//
+// - Parameters:
+//   - parent: The actual parent cell to be expanded
+    private func collapse(parent: ParentCell, _ tableView: UITableView) {
+        let numberOfChildren = dataSource.item(atRow: parent.index, inSection: parent.indexPath.section)?.children.count ?? 0
```

Comment:
Can we encapsulate `parent.indexPath.section` & `parent.index` to avoid repetition ?